Added DiVA fetcher based on the new interface

oscargus opened the pull request:
- [x] Tests created for changes

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,80 @@
+        try {
+            q = URLEncoder.encode(identifier, StandardCharsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            // this should never happen
```
Just throw a FetcherException as you do below.

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,80 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class DiVA implements IdBasedFetcher {
```
Can shortly explain what DiVA is in a comment and add a link to its API.

tobiasdiez replied:
I suggest to add this fetcher to the list of web searches and also remove the old one (or is there reason to keep the old code?)

oscargus replied:
As there is no easy way to wrap an IdBased searcher at the moment, it is
not possible. There was a class in the ISBN stupro PR so once that is merge
it makes sense.

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,80 @@
+            return Optional.empty();
+        }
+
+        String urlString = String.format(DiVA.URL_PATTERN, q);
```
Use Apache`s URIBuilder to create the URL.

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,80 @@
+    public HelpFile getHelpPage() {
+        return HelpFile.FETCHER_DIVA_TO_BIBTEX;
+    }
+
```
Could you please also add a method `isValidId(String identifier)` which checks that a given ID is of the expected format. I plan to add some functionality which depends on such a method but have no insights what DiVA identifiers look like. Thanks

oscargus replied:
I have no idea what the requirements are. Most I've seen are on the format `diva2:XXXXX` though, where X is a digit.

tobiasdiez replied:
I tried to play around with the API and it appears that the identifier definitely has to start with `diva2` so just take this as a test for valid identifier.

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,80 @@
+
+        Optional<BibEntry> result = Optional.ofNullable(BibtexParser.singleFromString(bibtexString));
+
+        result.ifPresent(entry -> entry.getFieldOptional(FieldName.INSTITUTION).ifPresent(institution -> entry
```
I would leave it to the users cleanup actions to format unicode to latex.

tobiasdiez commented on src/test/java/net/sf/jabref/logic/importer/fetcher/DiVATest.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Before
+    public void setUp() {
+        Globals.prefs = JabRefPreferences.getInstance();
```
Prefs are not required, right?

oscargus replied:
They are required to parse the BibTeX entry (wrapped fields).

oscargus commented:
@JabRef/developers: I rebased this and introduced a change in `MetaData`, making constructors without provided `Charset` deprecated. This is an attempt to get rid of the `Globals` dependency. Didn't change anything else related to that in this PR, but once merged there are many things to be changed.

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,69 @@
+            uriBuilder.addParameter("referenceFormat", "BibTex");
+            uriBuilder.addParameter("pids", identifier);
+
+            String bibtexString;
```
Combine this line with line 55.

oscargus replied:
Yeah, weird code... :-)

tobiasdiez commented on src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java:
```diff
@@ -0,0 +1,69 @@
+            URLDownload dl = new URLDownload(uriBuilder.build().toURL());
+
+            bibtexString = dl.downloadToString(StandardCharsets.UTF_8);
+            result = Optional.ofNullable(BibtexParser.singleFromString(bibtexString, importFormatPreferences));
```
Just `return Optional.of(...)` and remove result variable.

tobiasdiez commented:
Some small remarks. After fixing these and the codacy issue, this can be merged.
